tensorpipe/common/cuda_loop.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <condition_variable>
+#include <deque>
+#include <functional>
+#include <list>
+#include <mutex>
+#include <thread>
+
+#include <cuda_runtime.h>
+
+namespace tensorpipe {
+
+class CudaLoop {
+  using TCudaCallback = std::function<void(cudaError_t)>;
+
+  struct Operation {
+    TCudaCallback callback;
+    cudaError_t error;
+  };
+
+ public:
+  CudaLoop();
+
+  ~CudaLoop();
+
+  // TODO: device parameter?
+  void addCallback(cudaStream_t stream, TCudaCallback callback);
+
+ private:
+  std::thread thread_;
+  std::list<TCudaCallback> cudaCallbacks_;
```

Comment:
This list is used to store the callbacks in between the moment the user passes them to us and the moment CUDA tells us they're ready, right? So, during that time, "the ball is in CUDA's court", it's responsible for calling back into us so we can clean this. If we _really_ trusted CUDA we could just store the callback in a unique_ptr, pass it as the callbackPtr, "release" it on our side (AKA leak it), and then re-acquire it when CUDA calls into us.

I'm not saying to do so, as I'm not yet trusting CUDA that much, it's just a thought.